### PR TITLE
Bug 1829408: Reconcile machine power state annotation

### PR DESCRIPTION
**Pull request by alexander-demichev**
Reconcile machine power state annotation

### CODE REVIEW

File: pkg/controller/vsphere/reconciler.go [hunk 1]
```diff
@@ -333,6 +336,24 @@ func (r *Reconciler) reconcileNetwork(vm *virtualMachine) error {
 	return nil
 }
 
+func (r *Reconciler) reconcilePowerStateAnnontation(vm *virtualMachine) {
```
Comment (enxebre) [inline pkg/controller/vsphere/reconciler.go:339]: can we please have a unit for this func?

Comment (JoelSpeed) [reply pkg/controller/vsphere/reconciler.go:339]: +1, would be good to see a couple of tests exercising this explicitly

Comment (alexander-demichev) [reply pkg/controller/vsphere/reconciler.go:339]: done

File: pkg/controller/vsphere/reconciler.go [hunk 2]
```diff
@@ -333,6 +336,24 @@ func (r *Reconciler) reconcileNetwork(vm *virtualMachine) error {
...
+	if vm == nil {
+		return
+	}
+
+	// This can return an error if machine is being deleted
+	powerState, err := vm.getPowerState()
+	if err != nil {
+		klog.V(3).Infof("%s: Failed to get power state: %v", r.machine.Name, err)
+		return
```
Comment (enxebre) [inline pkg/controller/vsphere/reconciler.go:348]: why are we ignoring errors here?
we should return error, then is up to the consumer what to do with it, e.g log and ignore it if there's a reason to do so

File: pkg/controller/vsphere/reconciler_test.go [hunk 3]
```diff
@@ -1456,10 +1457,79 @@ func TestReconcileMachineWithCloudState(t *testing.T) {
 	}
 
 	if expectedProviderID != *reconciler.machine.Spec.ProviderID {
-		t.Errorf("Expected: %s, got: %s", expectedProviderID, *reconciler.machine.Spec.ProviderID)
+		t.Errorf("Expected providerId: %s, got: %s", expectedProviderID, *reconciler.machine.Spec.ProviderID)
+	}
+
+	expectedPowerState := "poweredOn"
```
Comment (enxebre) [inline pkg/controller/vsphere/reconciler_test.go:1463]: should this be a constant?

**Comment by alexander-demichev** [comment]
@enxebre All fixed

File: pkg/controller/vsphere/reconciler_test.go [hunk 4]
```diff
@@ -1456,10 +1459,87 @@ func TestReconcileMachineWithCloudState(t *testing.T) {
...
+	simulatorVM := simulator.Map.Any("VirtualMachine").(*simulator.VirtualMachine)
+	managedObjRef := simulatorVM.VirtualMachine.Reference()
+	vmObj := object.NewVirtualMachine(session.Client.Client, simulatorVM.Reference())
+	_, err := vmObj.PowerOn(context.TODO())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	vm := &virtualMachine{
+		Context: context.TODO(),
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/reconciler_test.go:1487]: Nit
```suggestion
		Context: context.Background(),
```

File: pkg/controller/vsphere/reconciler.go [hunk 5]
```diff
@@ -333,6 +338,25 @@ func (r *Reconciler) reconcileNetwork(vm *virtualMachine) error {
 	return nil
 }
 
+func (r *Reconciler) reconcilePowerStateAnnontation(vm *virtualMachine) error {
+	if vm == nil {
+		return errors.New("VM can't be nil")
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/reconciler.go:343]:
```suggestion
		return errors.New("provided VM is nil")
```

File: pkg/controller/vsphere/reconciler.go [hunk 6]
```diff
@@ -233,6 +233,11 @@ func (r *Reconciler) reconcileMachineWithCloudState(vm *virtualMachine, taskRef
 		return err
 	}
 
+	klog.V(3).Infof("%v: reconciling powerstate annotation", r.machine.GetName())
+	if err := r.reconcilePowerStateAnnontation(vm); err != nil {
+		klog.V(3).Infof("%s: Failed to get power state, machine can be in process of deletion: %v", r.machine.Name, err)
```
Comment (JoelSpeed) [inline pkg/controller/vsphere/reconciler.go:238]:
```suggestion
		klog.V(3).Infof("%s: Failed to get power state (machine may be being deleted): %v", r.machine.Name, err)
```

Comment (alexander-demichev) [reply pkg/controller/vsphere/reconciler.go]: done

File: pkg/controller/vsphere/reconciler.go (hunk 5 above)
Comment (alexander-demichev) [reply pkg/controller/vsphere/reconciler.go]: done

File: pkg/controller/vsphere/reconciler_test.go (hunk 4 above)
Comment (alexander-demichev) [reply pkg/controller/vsphere/reconciler_test.go]: 👍

File: pkg/controller/vsphere/reconciler.go [hunk 7]
```diff
@@ -233,6 +233,11 @@ func (r *Reconciler) reconcileMachineWithCloudState(vm *virtualMachine, taskRef
 		return err
 	}
 
+	klog.V(3).Infof("%v: reconciling powerstate annotation", r.machine.GetName())
+	if err := r.reconcilePowerStateAnnontation(vm); err != nil {
+		klog.V(3).Infof("%s: Failed to get power state (machine may be being deleted): %v", r.machine.Name, err)
```
Comment (enxebre) [inline pkg/controller/vsphere/reconciler.go:238]: why are we ignoring the error here? we should return an error.

Comment (enxebre) [reply pkg/controller/vsphere/reconciler.go:238]: shouldn't we?

**Pull request merged**